### PR TITLE
Add --warnings for deprecation logging

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

This is a first-step implementation of #3728 . More design spec please see https://docs.google.com/document/d/1x5pasO5z5wZWLJBKHbgVbEsSX60jSv9s3uJn8e8bwQc/edit

In short, we don't want users to see deprecation warnings by default. This PR suppresses all warnings by default and adds `--warnings` to control this behavior:

- No option. By default don't display all warnings and render a summary at the end of the build: "There're {} warnings in your build, please run --warnings=all to see them."
- `--warnings=no` Don't display warnings and summary.
- `--warnings=all` Just like what Gradle does now.

In addition, this PR refactors old `SingleMessageLogger` implementation and `LoggingDeprecation`. Previously, there're 3 parts scattered in `SingleMessageLogger`:

- Deprecation message construction
- Deprecation message logging
- Incubating message logging.

Now these parts are extracted into different `FeatureHandler` implementation to improve maintainability and readability:

- LoggingDeprecatedFeatureHandler
- LoggingIncubatingFeatureHandler

### CODE REVIEW

**Comment by blindpirate** [comment]
@oehme Could you take a look at it? I can find someone else if you're overburdened.

File: subprojects/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -24,6 +24,15 @@ Think of every feature section as a mini blog post.
   NOTE: Totally fine to just link to an example that show the feature.
 -->
 
+### Deprecation warning improvements
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: I recommend running prose through https://www.grammarly.com/, I find it's suggestions very useful.

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -24,6 +24,15 @@ Think of every feature section as a mini blog post.
...
+
+In this release, deprecation warnings are no longer displayed in build by default. Instead, all deprecation warnings in a build would be collected and a single warning would be rendered at the end of build:
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: ❌ will, not would

Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: ❌ "in the build log", not "in build"

File: subprojects/docs/src/docs/release/notes.md [hunk 3]
```diff
@@ -24,6 +24,15 @@ Think of every feature section as a mini blog post.
...
+
+    There're X deprecation warnings, which may break the build in Gradle 5.0. Please run with --warnings=all to see them.
+
+As said by this warning, you can run the build with command line option `--warnings=all` or property `org.gradle.warnings=all` to have all warnings displayed as before.
+Of course, you could use command line option `--warnings=no` or property `org.gradle.warnings=no` to suppress all warnings, including the one displayed at the end of build.
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: you can, not you could

File: subprojects/docs/src/docs/userguide/gradleCommandLine.adoc [hunk 4]
```diff
@@ -48,6 +48,14 @@ Set to `auto` (the default) to enable color and other rich output in the console
 Set to `rich` to enable color and other rich output in the console output, regardless of whether the build process is not attached to a console. When not attached to a console, the build output will use ANSI control characters to generate the rich output.
 +
 Set to `verbose` to enable color and other rich output like the `rich`, but output more detailed message (task name, output, etc.)
+`--warnings`::
+Specifies which type of deprecation warnings displayed in console.
++
+Set to `all` to render all deprecation warnings. This may be useful for users who're concern about these warnings.
++
+Set to `auto` (the default) to suppress all deprecation warnings in build but the "deprecation warning summary": `There're X deprecation warnings in your build.`
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/gradleCommandLine.adoc]: ❌ Should be named "summary"

File: subprojects/docs/src/docs/userguide/gradleCommandLine.adoc [hunk 5]
```diff
@@ -48,6 +48,14 @@ Set to `auto` (the default) to enable color and other rich output in the console
...
++
+Set to `no` to suppress all deprecation warnings as if they don't exist at all.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/gradleCommandLine.adoc]: ❌ Should be named "none"

File: subprojects/docs/src/docs/userguide/gradleCommandLine.adoc [hunk 6]
```diff
@@ -48,6 +48,14 @@ Set to `auto` (the default) to enable color and other rich output in the console
 Set to `rich` to enable color and other rich output in the console output, regardless of whether the build process is not attached to a console. When not attached to a console, the build output will use ANSI control characters to generate the rich output.
 +
 Set to `verbose` to enable color and other rich output like the `rich`, but output more detailed message (task name, output, etc.)
+`--warnings`::
+Specifies which type of deprecation warnings displayed in console.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/gradleCommandLine.adoc]: ❌ We should explicitly call them "Gradle deprecation warnings".  This setting does not affect compiler warnings and warnings from other tools.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/JULRedirector.java [hunk 7]
```diff
@@ -37,14 +38,9 @@
...
         super.start();
-        boolean shouldReadLoggingConfigFile = System.getProperty(READ_LOGGING_CONFIG_FILE_PROPERTY, "true").equals("true");
-        if (!shouldReadLoggingConfigFile) {
-            SingleMessageLogger.nagUserOfDiscontinuedProperty(READ_LOGGING_CONFIG_FILE_PROPERTY,
-                "Change your test to work with your java.util.logging configuration file settings.");
-        }
         if (!reset) {
             LogManager.getLogManager().reset();
-            if (shouldReadLoggingConfigFile) {
+            if (shouldReadLoggingConfigFile()) {
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/JULRedirector.java]: ❌ Why was this changed?

Comment (blindpirate) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/JULRedirector.java]: Before this change, deprecation check of `READ_LOGGING_CONFIG_FILE_PROPERTY` happened in test worker JVM, then the warning messages are forwarded to main JVM. It's impossible to collect warnings from another JVM, so we have to move this part to main JVM.

File: subprojects/docs/src/docs/release/notes.md (hunk 1 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Thank you, it's useful!

File: subprojects/docs/src/docs/userguide/gradleCommandLine.adoc (hunk 6 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/userguide/gradleCommandLine.adoc]: Personally I don't like to introduce too many build options. So here I think we can use `--warnings=no-deprecation` instead of an extra `--dep-warnings=none`. This allow us to extend the "warning type" in the future.

**Comment by blindpirate** [comment]
I've polished this PR slightly. @oehme @wolfs Maybe you'd like to take a look at it again.

File: subprojects/docs/src/docs/userguide/gradleCommandLine.adoc (hunk 6 above)
Comment (oehme) [reply subprojects/docs/src/docs/userguide/gradleCommandLine.adoc]: I wasn't talking about the name of the option, but its description. The description should clearly say these are *Gradle*  warnings, not Java compiler warnings etc. Unless we want to unify all of those.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Default value. By default, all deprecation warnings will be suppressed and a single summary message will be displayed at the end of a build indicating how many deprecation warnings suppressed.
+     */
+    Summary("summary"),
+
+    /**
+     * No deprecation warnings at all.
+     */
+    NoDeprecation("no-deprecation");
```
Comment (oehme) [inline subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java]: I'm not sure "no-deprecation" makes sense as a value of "warning type". Why not just set an empty set of warning types if I don't want to be notified of anything?

**Comment by blindpirate** [comment]
@oehme I'm really sorry for troubling you with the two parts together. Now I extract the warning part into #3847 so that it can be easier.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java (hunk 8 above)
Comment (blindpirate) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java]: That bothers me as well. If we make it as an empty set (or `None`), users will be confused because we only suppress deprecation warnings, not others (e.g.compiler warnings).

**Comment by blindpirate** [comment]
Superceded by #3847
